<a href="https://colab.research.google.com/github/reyonecy/ml_from_scratch1/blob/main/Customer_Call_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('/content/drive/MyDrive/Data_source/Customer Call List.xlsx')
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

## About Data
 - Given dataframe is a excel file which contains the  details about the costumers.
 - There are 8 features in the dataframe with following data type. They are :
     - CustomerID 	       (int64)
     - First_Name 	       (object)
     - Last_Name 	         (object)
     - Phone_Number 	     (object)   
     - Address 	           (object)
     - Paying Customer 	   (object)
     - Do_Not_Contact 	   (object)
     - Not_Useful_Column 	 (object)

- Customer ID is of data types int64 and    Not-Useful_Column is of bool and all other features are of type object
- There are 21 rows of data .


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   CustomerID         21 non-null     int64 
 1   First_Name         21 non-null     object
 2   Last_Name          20 non-null     object
 3   Phone_Number       19 non-null     object
 4   Address            21 non-null     object
 5   Paying Customer    21 non-null     object
 6   Do_Not_Contact     17 non-null     object
 7   Not_Useful_Column  21 non-null     bool  
dtypes: bool(1), int64(1), object(6)
memory usage: 1.3+ KB


In [4]:
df.dtypes

CustomerID            int64
First_Name           object
Last_Name            object
Phone_Number         object
Address              object
Paying Customer      object
Do_Not_Contact       object
Not_Useful_Column      bool
dtype: object

## Issues in Dataframe

1. Missing Data
  - Number of NaN in the follwing features.
      - Last_Name -	1
      - Phone_Number  -	2
      - Do_Not_Contact -	4

  -  Number of N/a in the follwing features.
      - Phone_Number - 	2
      - Address -	1
      - Paying Customer -	1


2. Duplicate data
  - Two same rows of data for the customer with Customer_ID 1020.
  - Two Same numbers 7066950392 in the Phone number  column.

3. Dirty Data
  - Last_Name has unwanted signs and symbols.
  - Phone_Number has inconsistent formating such as
   (/, |, etc) or no formating at all.
  - Paying Customer and	Do_Not_Contact has mixed up "Yes /No" and "Y/N".

4. Unwated feature called Not_useful_column is     present.  



In [5]:
df_null_count = df.isnull()
total_df_null_count = df_null_count.sum()
total_df_null_count

CustomerID           0
First_Name           0
Last_Name            1
Phone_Number         2
Address              0
Paying Customer      0
Do_Not_Contact       4
Not_Useful_Column    0
dtype: int64

In [6]:
Na_values = (df == "N/a").sum()
Na_values

CustomerID           0
First_Name           0
Last_Name            0
Phone_Number         2
Address              1
Paying Customer      1
Do_Not_Contact       0
Not_Useful_Column    0
dtype: int64

## Solutions Applied
 - Replaced missing values NaN and N/a with 'Unknown'
 - Removed duplicate data for Costumer_ID 1020
 - Removed unwanted leading and trailing signs and symbols in Last_Name  using str.strip()
 - Properly Formatted the phone numbers in Phone_Number column using a function().
 - Replaced Yes and No in the columns Paying Customer and Do_Not_Contact with 'Y' and 'N' respectively.
 -Removed the unwanted column called Not_Useful_Column.

In [7]:
df.fillna("Unknown",inplace=True)


In [8]:
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger       Unknown   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf      Unknown           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee       Unknown   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

In [9]:
df.replace("N/a","Unknown", inplace=True)
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger       Unknown   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf      Unknown       Unknown   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee       Unknown   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton       Unknown   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

In [10]:
df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20     True
dtype: bool

In [11]:
df.drop_duplicates(inplace=True)

In [12]:
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger       Unknown   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf      Unknown       Unknown   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee       Unknown   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton       Unknown   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [13]:
df['Last_Name'] = df['Last_Name'].str.strip('.!@#$%^&*()_-+=~`<>?/\\|[]{}"')


In [14]:
df


CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger       Unknown   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf     Unknown       Unknown   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee       Unknown   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton       Unknown   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [15]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger       Unknown   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf     Unknown       Unknown   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee       Unknown   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton       Unknown   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [16]:
import re

def format_phone_number(phone):
    # Check if the phone number is not an integer. If it is integer (Unknown) return it as is.
    if type(phone) == int:
      return phone

    # Otherwise, format it or return "Invalid Number"
    digits = re.sub(r'\D','',str(phone)) # Convert to string before using re.sub

    if len(digits) == 10:
        return f"{digits[:3]}-{digits[3:6]}-{digits[6:]}"
    else:
        return "Unknown" # Corrected: Return the string "Invalid Number"

df['Phone_Number'] = df['Phone_Number'].apply(format_phone_number)

In [17]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger       Unknown   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf     Unknown       Unknown   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee       Unknown   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton       Unknown   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes        Unknown   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [18]:
df['Paying Customer'] = df['Paying Customer'].replace({'Yes': 'Y', 'No': 'N'})
df['Do_Not_Contact'] = df['Do_Not_Contact'].replace({'Yes': 'Y', 'No': 'N'})

In [19]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger       Unknown   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf     Unknown       Unknown   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee       Unknown   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton       Unknown   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N        Unknown   
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y        Unknown   
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y        Unknown   
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y        Unknown   
18                                Unknown         Unknown              Y   
19            910 Tatooine Road, Tatooine               Y              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

In [21]:
df.drop('Not_Useful_Column', inplace=True, axis=1)

In [22]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger       Unknown   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf     Unknown       Unknown   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee       Unknown   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton       Unknown   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire               Y              N  
1                     93 West Main Street               N              Y  
2                      298 Drugs Driveway               N        Unknown  
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y  
4                        123 Dragons Road               Y              N  
5                        768 City Parkway               Y              Y  
6                       1209 South Street               N              N  
7                           98 Clue Drive               N              N  
8                        123 Middle Earth               Y        Unknown  
9              25th Main Street, New York               Y              N  
10                  612 Shire Lane, Shire               Y              N  
11                   2394 Hogwarts Avenue               Y        Unknown  
12                       2039 Main Street               Y              N  
13                       343 City Parkway               Y              N  
14                          214 HR Avenue               N              N  
15                   2395 Hogwarts Avenue               N              N  
16         121 Paper Avenue, Pennsylvania               Y              N  
17                        3498 Super Lane               Y        Unknown  
18                                Unknown         Unknown              Y  
19            910 Tatooine Road, Tatooine               Y              N

## Unsolved Issues
 - Couldn't figure out what to do with the duplicate phone number 7066950392 	in two rows with different costumers.
 - Also the function for formatting the phone number in correct way couldn't format those two duplicate phone numbers.

## Conclusion

The dataset, initially containing inconsistencies like missing values, duplicates, and formatting issues, has been cleaned.  Missing data was filled, duplicates removed, and phone numbers reformatted.  Remaining challenges include handling duplicate phone numbers associated with different customers and refining phone number formatting for specific cases.
